source/extensions/filters/network/kafka/kafka_request.h

```diff
@@ -0,0 +1,832 @@
+
+#include "common/common/assert.h"
+#include "envoy/common/exception.h"
+
+#include <sstream>
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+
+// === VECTOR ==================================================================
+
+template <typename T>
+std::ostream& operator<<(std::ostream &os, const std::vector<T>& arg) {
+  os << "[";
+  for (auto iter = arg.begin(); iter != arg.end(); iter++) {
+    if (iter != arg.begin()) os << ", ";
+    os << *iter;
+  }
+  os << "]";
+  return os;
+}
+
+template <typename T>
+std::ostream& operator<<(std::ostream &os, const absl::optional<T>& arg) {
+  if (arg.has_value()) {
+    os << *arg;
+  } else {
+    os << "<null>";
+  }
+  return os;
+}
+
+// === REQUEST =================================================================
+
+struct RequestHeader {
+  INT16 api_key_;
+  INT16 api_version_;
+  INT32 correlation_id_;
+  NULLABLE_STRING client_id_;
+
+  bool operator==(const RequestHeader& rhs) const {
+    return api_key_ == rhs.api_key_
+        && api_version_ == rhs.api_version_
+        && correlation_id_ == rhs.correlation_id_
+        && client_id_ == rhs.client_id_;
+  };
+
+  friend std::ostream& operator<<(std::ostream &os, const RequestHeader& arg) {
+    return os << "{api_key = " << arg.api_key_
+              << ", api_version = " << arg.api_version_
+              << ", correlation_id = " << arg.correlation_id_
+              << ", client_id = " << arg.client_id_
+              << "}";
+  };
+};
+
+struct RequestContext {
+  INT32 remaining_request_size_{0};
+  RequestHeader request_header_{};
+
+  friend std::ostream& operator<<(std::ostream &os, const RequestContext& arg) {
+    return os << "{header = " << arg.request_header_ << ", remaining = " << arg.remaining_request_size_ << "}";
+  }
+};
+
+typedef std::shared_ptr<RequestContext> RequestContextSharedPtr;
+
+// === REQUEST PARSER MAPPING (REQUEST TYPE => PARSER) =========================
+
+// a function generating a parser with given context
+typedef std::function<ParserSharedPtr(RequestContextSharedPtr)> GeneratorFunction;
+
+// two-level map: api_key -> api_version -> generator function
+typedef std::unordered_map<INT16, std::shared_ptr<std::unordered_map<INT16, GeneratorFunction>>> GeneratorMap;
+
+struct ParserSpec {
+  const INT16 api_key_;
+  const std::vector<INT16> api_versions_;
+  const GeneratorFunction generator_;
+};
+
+// helper function that generates a map from specs looking like { api_key, api_versions... }
+GeneratorMap computeGeneratorMap(std::vector<ParserSpec> arg);
+
+/**
+ * Responsible for mapping [apiKey, apiVersion] -> request parser
+ */
+class RequestParserResolver {
+public:
+  RequestParserResolver(std::vector<ParserSpec> arg): generators_{computeGeneratorMap(arg)} {};
+
+  ParserSharedPtr createParser(INT16 api_key, INT16 api_version, RequestContextSharedPtr context_) const;
+
+  static const RequestParserResolver KAFKA_0_11;
```

Comment:
left that separation explicitly, I'd intend to have various versions such as `KAFKA_10`, `KAFKA_11` etc. depending on configuration - this might allow us to e.g. perform request rewrites when someone sends a request that cannot be handled by cluster (or the other way?)